shared/src/main/scala/org/adridadou/openlaw/parser/template/OpenlawTemplateLanguageParserService.scala

```diff
@@ -8,49 +8,43 @@ import org.adridadou.openlaw.parser.template.printers.XHtmlAgreementPrinter.Frag
 import org.adridadou.openlaw.parser.template.variableTypes.YesNoType
 import org.parboiled2.{ErrorFormatter, ParseError}
 
-import scala.util.{Failure, Success, Try}
 import cats.implicits._
 import org.adridadou.openlaw.parser.template.expressions.Expression
+import org.adridadou.openlaw.result.{attempt, Failure, handleFatalErrors, Result, Success}
 
 /**
   * Created by davidroon on 05.06.17.
   */
 class OpenlawTemplateLanguageParserService(val internalClock:Clock) {
 
-  def parseExpression(str: String): Either[String, Expression] = new ExpressionParser(str).root.run() match {
-    case Success(result) => Right(result)
-    case Failure(ex) => Left(ex.getMessage)
+  def parseExpression(str: String): Result[Expression] = new ExpressionParser(str).root.run() match {
+    case scala.util.Success(result) => Success(result)
+    case scala.util.Failure(ex) => handleFatalErrors(ex)
   }
 
+  /*
   def compileTemplateOrThrow(content: String): CompiledTemplate = compileTemplate(content) match {
     case Right(document) => document
-    case Left(ex) => throw new RuntimeException(ex)
+    case Left(ex) => Failure(ex)
   }
+  */
 
-  def compileTemplate(source: String, clock: Clock = internalClock): Either[String, CompiledTemplate] = {
+  def compileTemplate(source: String, clock: Clock = internalClock): Result[CompiledTemplate] = {
     val compiler = createTemplateCompiler(source, clock)
 
-    Try(compiler.rootRule.run().toEither match {
-      case Left(parseError: ParseError) =>
-        Left(compiler.formatError(parseError))
-      case Left(ex) =>
-        Left(Option(ex.getMessage).getOrElse(""))
-      case Right(result) =>
-        validate(result)
-    }) match {
-      case Success(result) =>
-        result
-      case Failure(ex) =>
-        Left(Option(ex.getMessage).getOrElse(""))
+    compiler.rootRule.run() match {
```

Comment:
the issue here is that we throw exceptions within the parser. So if you don't use Try here, you may end up with an unhandled exception

Reply:
maybe use attempt here

Reply:
Yes, good catch.  Missed this one.

Reply:
The parser is a particularly odd API in that it returns a Try but also seems to throw exceptions outside of it.